ykllvmwrap/src/jitmodbuilder.cc

```diff
@@ -121,14 +121,28 @@ class FuncAddrs {
 /// Extract all live variables that need to be passed into the control point.
 /// FIXME: This is currently an overapproximation and will return some
 /// variables that are no longer alive.
-std::vector<Value *> getLiveVars(DominatorTree &DT, Instruction *Before) {
+std::vector<Value *> getLiveVars(
+    DominatorTree &DT, Instruction *Before,
+    std::map<Value *, std::tuple<size_t, size_t, Instruction *, size_t>> AOTMap,
+    size_t CallDepth) {
   std::vector<Value *> Vec;
   Function *Func = Before->getFunction();
   for (auto &BB : *Func) {
     if (!DT.dominates(cast<Instruction>(Before), &BB)) {
       for (auto &I : BB) {
         if ((!I.getType()->isVoidTy()) &&
             (DT.dominates(&I, cast<Instruction>(Before)))) {
+          std::tuple<size_t, size_t, Instruction *, size_t> Entry = AOTMap[&I];
+          size_t StackFrameIdx = std::get<3>(Entry);
+          if (StackFrameIdx > CallDepth) {
+            // Variables from a previously inlined function call are no longer
```

Comment:
If possible I'd also like to see a test which tickles this code and is currently marked "ignored" so that we know to come back and fix it.

Reply:
Fixed in 78bf9b44f.

There already are tests that tickle this. I had to add this `if` to fix them as they would fail otherwise. There's nothing to fix here really. Once we have implemented proper liveness we can remove this whole function.

Reply:
The comment makes it sound like a correctness fix though?

Reply:
I'm not sure. The current "liveness analysis" is picking up more live variables than it needs to at each guard failure. That is not necessarily incorrect and just mean we are initialising more variables in the stopgap interpreter than we have to. With this PR, however, this means that the extra variables the liveness picks up, may come from the frame of a function call that is already finished (at the time of a guard failure). Trying to add those to the stopgap interpreter thus results in an error (as the stackframe index of such variables is higher than there are active frames in the stopgap interpreter). The above code fixes this by omitting live variables whose stackframe is already finished.

In other words, the current code is not incorrect, but contains a fix that will likely not be needed any more once we have proper liveness analysis in `jitmodbuilder`.

Reply:
I think I understand: can you maybe try rephrasing the comment to bring the above out more clearly?

Reply:
Something along the lines of:

> Our liveness analysis over-approximates the live set and includes values which cannot be mapped back... Until we have precise liveness, we filter out the extra values in an ad-hoc fashion.

?